cpp/src/join/distinct_hash_join.cu

```diff
@@ -157,7 +156,8 @@ distinct_hash_join::inner_join(cudf::table_view const& probe,
     std::make_unique<rmm::device_uvector<size_type>>(probe_table_num_rows, stream, mr);
 
   auto const probe_row_hasher = cudf::experimental::row::hash::row_hasher{preprocessed_probe};
-  auto const d_probe_hasher   = probe_row_hasher.device_hasher(nullate::DYNAMIC{this->_has_nulls});
+  auto const has_nulls        = _has_nulls or cudf::has_nested_nulls(probe);
+  auto const d_probe_hasher   = probe_row_hasher.device_hasher(nullate::DYNAMIC{has_nulls});
```

Comment:
I still see problem here. When `has_nulls` is `true` due to nulls in the probe table but `_has_nulls` is `false`, the `build` table was hashed by a different way than the probe table. As such, two identical rows in two tables may be hashed to different values, which lead to incorrect join output.

Reply:
I finally get what you mean. Should we do the same cleanup for `hash_join` as well?

Reply:
Oh so the same issue was already merged before? I didn't know that. Yes, I think this is a real bug and should be addressed completely. If that issue was already merged before and is not related to this PR, we can fix it in a separate work.

Reply:
distinct hash join is fixed in this PR. I will address `hash_join` separately.